src/Razor/test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Cohost/CohostEndpointTest.cs

```diff
@@ -0,0 +1,186 @@
+        Assert.Equal(actualProviders.OrderBy(a => a.Name).Select(r => r.Name).ToArray(), providers.OrderBy(e => e.GetType().Name).Select(r => r.GetType().Name).ToArray());
+
+        var clientCapabilities = new VSInternalClientCapabilities()
+        {
+            SupportsVisualStudioExtensions = true,
+            TextDocument = new TextDocumentClientCapabilities()
+            {
+                CodeAction = new() { DynamicRegistration = true },
+                CodeLens = new() { DynamicRegistration = true },
+                Completion = new() { DynamicRegistration = true },
+                Definition = new() { DynamicRegistration = true },
+                Diagnostic = new() { DynamicRegistration = true },
+                DocumentHighlight = new() { DynamicRegistration = true },
+                DocumentLink = new() { DynamicRegistration = true },
+                DocumentSymbol = new() { DynamicRegistration = true },
+                FoldingRange = new() { DynamicRegistration = true },
+                Formatting = new() { DynamicRegistration = true },
+                Hover = new() { DynamicRegistration = true },
+                Implementation = new() { DynamicRegistration = true },
+                InlayHint = new() { DynamicRegistration = true },
+                LinkedEditingRange = new() { DynamicRegistration = true },
+                OnTypeFormatting = new() { DynamicRegistration = true },
+                RangeFormatting = new() { DynamicRegistration = true },
+                References = new() { DynamicRegistration = true },
+                Rename = new() { DynamicRegistration = true },
+                SemanticTokens = new() { DynamicRegistration = true },
+                SignatureHelp = new() { DynamicRegistration = true },
+                Synchronization = new() { DynamicRegistration = true },
+                TypeDefinition = new() { DynamicRegistration = true }
+            },
+        };
+
+        foreach (var endpoint in providers)
+        {
+            if (endpoint is CohostSemanticTokensRangeEndpoint)
+            {
+                // We can't currently test this, as the GetRegistrations method calls requestContext.GetRequiredService
+                // and we can't create a request context ourselves
+                continue;
+            }
+
+            var registrations = endpoint.GetRegistrations(clientCapabilities, requestContext: new());
+
+            // If we didn't get any registrations then the test is probably invalid, and we need to update client capabilities above
+            if (registrations.Length == 0)
+            {
+                Assert.Fail($"Did not get any registrations from {endpoint.GetType().Name}. Client capabilities might be wrong?");
+            }
+
+            foreach (var registration in registrations)
+            {
+                var options = registration.RegisterOptions as ITextDocumentRegistrationOptions;
```

Comment:
💭I'm not sure what the error message looks like but there is `Assert.IsType<T>`

Reply:
Tyme to start writing extension methods for ALL of XUnit :P

Reply:
Not sure that'd work out, since xUnit's `Assert` is a static class and we don't have the ability to create static extension methods. IMO, we'd be better off just using FluentAssertions or some other library.

Reply:
Or do the `AssertEx` thing for our own implementations